### PR TITLE
3 New Strategies From the PRISON (http://www.lifl.fr/IPD/ipd.frame.html)

### PR DESCRIPTION
**Pull request by geraintpalmer**
EasyGo
GrudgerAlternator
GradualKiller

### CODE REVIEW

**Comment by geraintpalmer** [comment]
Hi! Just looked at Issue  #379 and implemented 3 new strategies. Hope this helps 😄

File: axelrod/strategies/gradualkiller.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from axelrod import Actions, Player, init_args
+C, D = Actions.C, Actions.D
+
+
+class GradualKiller(Player):
+    """
+    It begins by defecting in the first five moves, then cooperates two times.
+    It then defects all the time if the opponent has defected in move 6 and 7,
+    else cooperates all the time.
```
Comment (drvinceknight) [inline axelrod/strategies/gradualkiller.py]: Could you add a sentence here saying that this is designed to stop `Gradual` from winning? (And possibly why?)

File: axelrod/strategies/gradualkiller.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    def strategy(self, opponent):
+        """This is the actual strategy"""
+        # First seven moves
+        firstseven = [D, D, D, D, D, C, C]
+        if len(self.history) < 7:
+            return firstseven[len(self.history)]
+        # React to the opponent's 6th and 7th moves
+        elif opponent.history[5:7] == [D, D]:
```
Comment (drvinceknight) [inline axelrod/strategies/gradualkiller.py]: This can just be an `if` statement.

File: axelrod/strategies/grudger.py [hunk 3]
```diff
@@ -171,3 +171,62 @@ def reset(self):
         Player.reset(self)
         self.grudged = False
         self.grudge_memory = 0
+
+
+
+class GrudgerAlternator(Player):
+    """
+    A player starts by cooperating until the first opponents defenction,
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: "defection"

File: axelrod/strategies/gradualkiller.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def strategy(self, opponent):
+        """This is the actual strategy"""
+        # First seven moves
+        firstseven = [D, D, D, D, D, C, C]
```
Comment (meatballs) [inline axelrod/strategies/gradualkiller.py]: We have the ```@initial_sequence``` decorator to handle this

File: axelrod/strategies/grudger.py [hunk 5]
```diff
@@ -171,3 +171,62 @@ def reset(self):
...
+    then alternates D-C.
+
+    Names:
+
+    - c_then_per_dc: [PRISON1998]_
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: Can you also add `GrudgerAlternator` to this list?

File: axelrod/strategies/grudger.py [hunk 6]
```diff
@@ -171,3 +171,62 @@ def reset(self):
...
+        if opponent.defections:
+            if self.history[-1] == Actions.C:
+                return Actions.D
+        return Actions.C
+
+
+
+class EasyGo(Player):
+    """
+    A player starts by defecting however will cooperate if at any point the opponent has defected.
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: if the opponent

File: docs/reference/bibliography.rst [hunk 7]
```diff
@@ -18,3 +18,4 @@ documentation.
 .. [Stewart2012] Stewart, a. J., & Plotkin, J. B. (2012). Extortion and cooperation in the Prisoner’s Dilemma. Proceedings of the National Academy of Sciences, 109(26), 10134–10135. http://doi.org/10.1073/pnas.1208087109
 .. [Szabó1992] Szabó, G., & Fáth, G. (2007). Evolutionary games on graphs. Physics Reports, 446(4-6), 97–216. http://doi.org/10.1016/j.physrep.2007.04.004
 .. [Tzafestas2000] Tzafestas, E. (2000). Toward adaptive cooperative behavior. From Animals to Animals: Proceedings of the 6th International Conference on the Simulation of Adaptive Behavior {(SAB-2000)}, 2, 334–340.
+.. [PRISON1998] LIFL (1998) PRISON. Available at: http://www.lifl.fr/IPD/ipd.frame.html (Accessed: 19 September 2016).
```
Comment (drvinceknight) [inline docs/reference/bibliography.rst]: Is there no academic reference?

File: axelrod/tests/unit/test_grudger.py [hunk 8]
```diff
@@ -173,3 +174,86 @@ def test_reset(self):
...
+        # Defected at least once
+        for _ in range(20):
+            i = randint(1, 30)
+            j = randint(1, 30)
+            opp_hist = [C for r in range(i)] + [D]
+            my_hist = [C] * (i + 1)
+            expected_response = [D if r % 2 == 0 else C for r in range(j)]
+            self.responses_test(my_hist, opp_hist, expected_response)
+
+
```
Comment (drvinceknight) [inline axelrod/tests/unit/test_grudger.py]: Can you get rid of the extra lines? Should be 2.

File: axelrod/strategies/gradualkiller.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        # # First seven moves
+        # firstseven = [D, D, D, D, D, C, C]
+        # if len(self.history) < 7:
+        #     return firstseven[len(self.history)]
+        # # React to the opponent's 6th and 7th moves
+        # elif opponent.history[5:7] == [D, D]:
+        #     return D
+        # return C
+
+        if len(self.history) >= 7:
```
Comment (meatballs) [inline axelrod/strategies/gradualkiller.py]: I think this condition is redundant now that the decorator is being used

Comment (geraintpalmer) [inline axelrod/strategies/gradualkiller.py]: Wouldn't this line:

```
if opponent.history[5:7] == [D, D]:
```

not break before the 7th round?

File: axelrod/strategies/grudger.py (hunk 6 above)
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: Ignore this.

File: axelrod/strategies/gradualkiller.py (hunk 9 above)
Comment (meatballs) [inline axelrod/strategies/gradualkiller.py]: The decorator should override it in the early turns. Try it!

Comment (geraintpalmer) [inline axelrod/strategies/gradualkiller.py]: You're right! 👍

Comment (meatballs) [inline axelrod/strategies/gradualkiller.py]: I try to be once a month or so!

File: axelrod/strategies/gradualkiller.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        'stochastic': False,
+        'makes_use_of': set(),
+        'long_run_time': False,
+        'inspects_source': False,
+        'manipulates_source': False,
+        'manipulates_state': False
+    }
+
+    def strategy(self, opponent):
+        """This is the actual strategy"""
```
Comment (marcharper) [inline axelrod/strategies/gradualkiller.py]: Please delete this line and the commented lines below.

File: axelrod/strategies/grudger.py [hunk 11]
```diff
@@ -171,3 +171,63 @@ def reset(self):
...
+
+class GrudgerAlternator(Player):
+    """
+    A player starts by cooperating until the first opponents defection,
+    then alternates D-C.
+
+    Names:
+
+    - c_then_per_dc: [PRISON1998]_
+    - GrudgerAlternator
```
Comment (drvinceknight) [inline axelrod/strategies/grudger.py]: Can you make this something like:

```
- Grudger Alternator: Original name by Geraint Palmer
```

Or whatever self attribution you would like to use?

**Pull request merged**